irs/core/upload_handler.py

```diff
@@ -0,0 +1,235 @@
+            total_sha256_checksum.update(decrypted)
+
+    # process remaining data
+    if incomplete_chunk:
+        decrypted = decrypt_block(ciphersegment=incomplete_chunk, session_keys=[secret])
+        total_sha256_checksum.update(decrypted)
+
+    return (
+        encrypted_md5_part_checksums,
+        encrypted_sha256_part_checksums,
+    ), total_sha256_checksum.hexdigest()
+
+
+async def retrieve_parts(*, url: str, object_size: int, offset: int = 0):
+    """Get all parts from inbox, starting with file content at offset"""
+    for start, stop in calc_part_ranges(
+        part_size=PART_SIZE, object_size=object_size, byte_offset=offset
+    ):
+        yield retrieve_part(url=url, start=start, stop=stop)
+
+
+async def retrieve_part(*, url: str, start: int, stop: int):
+    """Get one part from inbox by range"""
+    try:
+        response = requests.get(
+            url=url, headers={"Range": f"bytes={start}-{stop}"}, timeout=60
+        )
+    except requests.exceptions.RequestException as request_error:
+        raise exceptions.RequestFailedError(url=url) from request_error
+
+    return response.content
+
+
+def calc_part_ranges(
+    *, part_size: int, object_size: int, byte_offset: int
+) -> Sequence[tuple[int, int]]:
+    """
+    Calculate and return the ranges (start, end) of file parts as a list of tuples.
+    """
+    # calc the ranges for the parts that have the full part_size:
+    full_part_number = math.floor(object_size / part_size)
+
+    part_ranges = [
+        (
+            byte_offset + part_size * (part_no - 1),
+            byte_offset + part_size * (part_no) - 1,
+        )
+        for part_no in range(1, full_part_number + 1)
+    ]
+
+    if (object_size % part_size) > 0:
+        # if the last part is smaller than the part_size, calculate it range separately:
+        part_ranges.append(
+            (byte_offset + part_size * full_part_number, object_size - 1)
+        )
+    return part_ranges
+
+
+def make_chunks(*, file_part: bytes) -> Tuple[List[bytes], bytes]:
+    """Chunk file part into decryptable chunks"""
+
+    num_chunks = len(file_part) / CIPHER_SEGMENT_SIZE
+    full_chunks = int(num_chunks)
+    print(num_chunks)
+    chunks = [
+        file_part[i * CIPHER_SEGMENT_SIZE : (i + 1) * CIPHER_SEGMENT_SIZE]
+        for i in range(full_chunks)
+    ]
+
+    # check if we have a remainder of bytes that we need to handle,
+    # i.e. non-matching boundaries between part and cipher segment size
+    incomplete_chunk = b""
+    partial_chunk_idx = math.ceil(num_chunks)
+    if partial_chunk_idx != full_chunks:
+        incomplete_chunk = file_part[full_chunks * CIPHER_SEGMENT_SIZE :]
+    return chunks, incomplete_chunk
+
+
+def get_part_checksums(*, file_part: bytes):
+    """Compute md5 and sha256 for encrypted part"""
+    return (
+        hashlib.md5(file_part, usedforsecurity=False).hexdigest(),
+        hashlib.sha256(file_part).hexdigest(),
+    )
```

Comment:
The core.models file is empty, can be removed. (There was no way to directly annotate that file.)